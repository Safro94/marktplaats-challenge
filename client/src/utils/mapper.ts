import { IBrand, IModel } from '../types';

const mapDropdownOptions = (data: IBrand[] | IModel[]) =>
	data.map((item: IBrand | IModel) => ({
		label: item.name,
		value: item.id.toString(),
	}));

export { mapDropdownOptions };
