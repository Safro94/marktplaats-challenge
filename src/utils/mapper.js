const mapBrands = brands =>
	brands.map(({ id, name }) => ({
		id,
		name,
	}));

module.exports = { mapBrands };
