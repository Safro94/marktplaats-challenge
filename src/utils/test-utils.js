const mockReq = params => {
	const req = { ...params };
	return req;
};

const mockRes = params => {
	const res = { ...params };
	res.json = jest.fn().mockReturnValue(res);
	res.status = jest.fn().mockReturnValue(res);

	return res;
};

const mappedBrands = [
	{
		id: 1,
		name: 'Honda',
	},
	{
		id: 2,
		name: 'Ferrari',
	},
];

module.exports = {
	mockReq,
	mockRes,
	mappedBrands,
};
