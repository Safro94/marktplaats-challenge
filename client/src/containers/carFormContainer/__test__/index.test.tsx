import userEvent from '@testing-library/user-event';
import { render, screen, waitFor } from '../../../utils/test-utils';

import CarFormContainer from '../';

import BrandService from '../../../services/brandService';
import ModelService from '../../../services/modelService';

jest.mock('../../../services/brandService', () => {
	return {
		getBrands: jest.fn(),
	};
});

jest.mock('../../../services/modelService', () => {
	return {
		getModelByBrandId: jest.fn(),
	};
});

const mockHandleError = jest.fn();
jest.mock('react-error-boundary', () => ({
	...jest.requireActual('react-error-boundary'),
	useErrorHandler: () => mockHandleError,
}));

describe('CarFormContainer', () => {
	const brandId = 1;
	let origErrorConsole: {
		(...data: any[]): void;
		(message?: any, ...optionalParams: any[]): void;
		(...data: any[]): void;
		(message?: any, ...optionalParams: any[]): void;
	};

	beforeEach(() => {
		(BrandService.getBrands as jest.Mock).mockResolvedValue([
			{ name: 'ferrari', id: 1 },
		]);

		(ModelService.getModelByBrandId as jest.Mock).mockResolvedValue([
			{ name: 'F8', id: 1 },
			{ name: '458 spider', id: 2 },
		]);

		origErrorConsole = window.console.error;

		window.console.error = (...args) => {
			const firstArg = args.length > 0 && args[0];

			const shouldBeIgnored =
				firstArg &&
				typeof firstArg === 'string' &&
				firstArg.includes('Not implemented: HTMLFormElement.prototype.submit');

			if (!shouldBeIgnored) {
				origErrorConsole(...args);
			}
		};
	});

	afterEach(() => {
		window.console.error = origErrorConsole;
	});

	it('should render correctly', async () => {
		const { container } = render(<CarFormContainer />);

		await waitFor(() => {
			expect(container.firstChild).toMatchSnapshot();
		});
	});

	it('should show a disabled button', async () => {
		render(<CarFormContainer />);

		await waitFor(() => {
			expect(
				screen.getByRole('button', { name: /search car/i })
			).toBeDisabled();
		});
	});

	it('should enable the button when keywords is not empty', async () => {
		render(<CarFormContainer />);

		const searchButton = screen.getByRole('button', {
			name: /search car/i,
		});
		expect(searchButton).toBeDisabled();

		const keywordsInput = screen.getByPlaceholderText(/keywords/i);
		userEvent.type(keywordsInput, 'a');

		await waitFor(() => {
			expect(searchButton).not.toBeDisabled();
		});
	});

	it('should fill the models select once the brand is selected and call the services', async () => {
		render(<CarFormContainer />);

		const searchButton = screen.getByRole('button', { name: /search car/i });
		expect(searchButton).toBeDisabled();

		await waitFor(() => {
			expect(BrandService.getBrands).toHaveBeenCalledTimes(1);
			expect(BrandService.getBrands).toHaveBeenCalledWith(mockHandleError);
		});

		const brandSelectButton = screen.getAllByTestId('toggle-button')[0];
		userEvent.click(brandSelectButton);

		const selectedOption = screen.getByRole('option', { name: /ferrari/i });
		userEvent.click(selectedOption);
		expect(searchButton).not.toBeDisabled();

		await waitFor(() => {
			expect(ModelService.getModelByBrandId).toHaveBeenCalledTimes(1);
			expect(ModelService.getModelByBrandId).toHaveBeenCalledWith(
				brandId,
				mockHandleError
			);
		});
	});

	it('should change the selected item when the user selects another model', async () => {
		render(<CarFormContainer />);

		await waitFor(() => {
			expect(BrandService.getBrands).toHaveBeenCalledTimes(1);
			expect(BrandService.getBrands).toHaveBeenCalledWith(mockHandleError);
		});

		const brandSelectButton = screen.getAllByTestId('toggle-button')[0];
		userEvent.click(brandSelectButton);

		const selectedBrand = screen.getByRole('option', { name: /ferrari/i });
		userEvent.click(selectedBrand);

		const ModelSelectButton = screen.getAllByTestId('toggle-button')[1];
		expect(ModelSelectButton).toHaveTextContent(/select a model/i);

		await waitFor(() => {
			expect(ModelSelectButton).toHaveTextContent(/F8/i);
			userEvent.click(ModelSelectButton);
		});

		const newSelectedModel = screen.getByRole('option', {
			name: /458 spider/i,
		});
		userEvent.click(newSelectedModel);
		expect(ModelSelectButton).toHaveTextContent(/458 spider/i);
	});

	it('should call the submit function when the button is clicked', async () => {
		window.HTMLFormElement.prototype.submit = () => {};
		const logSpy = jest.spyOn(console, 'log').mockImplementation(() => {});

		render(<CarFormContainer />);

		const searchButton = screen.getByRole('button', {
			name: /search car/i,
		});

		const keywordsInput = screen.getByPlaceholderText(/keywords/i);
		userEvent.type(keywordsInput, 'a');

		await waitFor(() => {
			userEvent.click(searchButton);
		});

		expect(logSpy).toBeCalledTimes(1);
		expect(logSpy).toBeCalledWith('submit');
	});
});
