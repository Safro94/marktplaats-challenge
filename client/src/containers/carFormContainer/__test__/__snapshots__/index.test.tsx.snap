// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarFormContainer should render correctly 1`] = `
<div
  class="sc-crrszt egaUpg"
>
  <form
    class="sc-fubCzh hXKFlP"
    method="POST"
    rows="3"
  >
    <div
      class="sc-pGacB fEGoAx"
    >
      <h2
        class="sc-jrAFXE dzOuyr"
      >
        Buy a car
      </h2>
    </div>
    <div
      class="sc-pGacB fEGoAx"
    >
      <div
        class="sc-dQoVA YydjD"
      >
        <div
          class="sc-iBPTik fzbgYv"
        >
          <label
            class="sc-kEjbQP bfVRwd"
            for="S1"
            id="downshift-0-label"
          >
            Brand:
          </label>
        </div>
        <div
          class="sc-gsTEea lheJaR"
        >
          <div
            class="sc-dlfnuX fTBBZX"
          >
            <svg
              class="sc-hKgJUU jsSjCx"
            >
              chevron-down.svg
            </svg>
          </div>
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="downshift-0-label downshift-0-toggle-button"
            class="sc-eCstlR dohvVr"
            data-testid="toggle-button"
            id="S1"
            type="button"
          >
            Select a brand
          </button>
          <ul
            aria-labelledby="downshift-0-label"
            class="sc-jSgvzq fyJcCp"
            id="downshift-0-menu"
            role="listbox"
            tabindex="-1"
          />
        </div>
      </div>
      <div
        class="sc-dQoVA YydjD"
      >
        <div
          class="sc-iBPTik fzbgYv"
        >
          <label
            class="sc-kEjbQP bfVRwd"
            for="S2"
            id="downshift-1-label"
          >
            Model:
          </label>
        </div>
        <div
          class="sc-gsTEea lheJaR"
        >
          <div
            class="sc-dlfnuX fTBBZX"
          >
            <svg
              class="sc-hKgJUU jsSjCx"
            >
              chevron-down.svg
            </svg>
          </div>
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="downshift-1-label downshift-1-toggle-button"
            class="sc-eCstlR kbrHkS"
            data-testid="toggle-button"
            disabled=""
            id="S2"
            type="button"
          >
            Select a model
          </button>
          <ul
            aria-labelledby="downshift-1-label"
            class="sc-jSgvzq fyJcCp"
            id="downshift-1-menu"
            role="listbox"
            tabindex="-1"
          />
        </div>
      </div>
      <div
        class="sc-dQoVA YydjD"
      >
        <label
          class="sc-kEjbQP bfVRwd"
          for="T"
        >
          Keywords:
        </label>
        <input
          class="sc-iqHYmW gZgkqK"
          id="T"
          name="keywords"
          placeholder="Keywords"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-pGacB fEGoAx"
    >
      <button
        class="sc-bdfBQB bLzkoJ"
        disabled=""
        id="B"
        type="submit"
      >
        Search Cars
      </button>
    </div>
  </form>
</div>
`;
